File: src/processors/AutoRotate.py
from pathlib import Path

import cv2

from src.processors.interfaces.ImageTemplatePreprocessor import (
    ImageTemplatePreprocessor,
)
from src.utils.image import ImageUtils
from src.utils.interaction import InteractionUtils
from src.utils.logger import logger


class AutoRotate(ImageTemplatePreprocessor):
    __is_internal_preprocessor__ = False

    def get_class_name(self) -> str:
        return "AutoRotate"

    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        path = self.get_relative_path(self.options["referenceImage"])
        if not path.exists():
            msg = f"Reference image for AutoRotate not found at {path}"
            raise Exception(msg)
        self.reference_image = cv2.imread(str(path), cv2.IMREAD_GRAYSCALE)
        self.marker_dimensions = self.options.get("markerDimensions", None)
        self.resized_reference = self.reference_image
        self.threshold = self.options.get("threshold", None)
        if self.marker_dimensions:
            self.resized_reference = ImageUtils.resize_to_dimensions(
                self.marker_dimensions, self.reference_image
            )

    def apply_filter(self, image, colored_image, _template, _file_path):
        config = self.tuning_config
        best_val, best_rotation = -1, None
        rotations = [
            None,
            cv2.ROTATE_90_CLOCKWISE,
            cv2.ROTATE_180,
            cv2.ROTATE_90_COUNTERCLOCKWISE,
        ]
        for rotation in rotations:
            rotated_img = image
            if rotation is not None:
                rotated_img = ImageUtils.rotate(
                    image, rotation, keep_original_shape=True
                )
            # TODO: find a better suited template matching for white images.
            res = cv2.matchTemplate(
                rotated_img, self.resized_reference, cv2.TM_CCOEFF_NORMED
            )
            _min_val, max_val, _min_loc, _max_loc = cv2.minMaxLoc(res)

            if config.outputs.show_image_level >= 5:
                hstack = ImageUtils.get_padded_hstack(
                    [rotated_img, self.resized_reference]
                )
                InteractionUtils.show("Template matching result", res, 0)
                InteractionUtils.show(
                    f"Template Matching for rotation: {'No Rotation' if rotation is None else 90 * (1 + rotation)} ({max_val:.2f})",
                    hstack,
                )

            if max_val > best_val:
                best_val = max_val
                best_rotation = rotation

        if self.threshold is not None and self.threshold["value"] > best_val:
            if self.threshold["passthrough"]:
                logger.warning(
                    "The autorotate score is below threshold. Continuing due to passthrough flag."
                )
            else:
                logger.error(
                    "The autorotate score is below threshold. Adjust your threshold or check the reference marker and input image."
                )
                msg = "The autorotate score is below threshold"
                raise Exception(msg)

        logger.info(
            f"AutoRotate Applied with rotation {best_rotation} and value {best_val}"
        )
        if best_rotation is None:
            return image, colored_image, _template

        image = ImageUtils.rotate(image, best_rotation, keep_original_shape=True)

        if config.outputs.show_image_level >= 4:
            InteractionUtils.show(
                f"Image after AutoRotate: rotation - {best_rotation}",
                image,
                1,
            )
        if self.tuning_config.outputs.colored_outputs_enabled:
            colored_image = ImageUtils.rotate(
                colored_image, best_rotation, keep_original_shape=True
            )
        return image, colored_image, _template

    def exclude_files(self) -> list[Path]:
        path = self.get_relative_path(self.options["referenceImage"])
        return [path]

File: omrchecker-js/packages/core/src/processors/image/AutoRotate.ts
/**
 * Migrated from Python: src/processors/image/AutoRotate.py
 * Agent: Oz
 * Phase: Filter Processors
 *
 * Automatic rotation detection and correction using template matching.
 */

import cv from '@techstark/opencv-js';
import { ProcessingContext, Processor } from '../base';

export interface AutoRotateOptions {
  referenceImage: cv.Mat;
  markerDimensions?: [number, number];
  threshold?: {
    value: number;
    passthrough: boolean;
  };
}

/**
 * Automatically detects and corrects document orientation.
 * 
 * Uses template matching to compare the input image against a reference marker
 * in all four rotations (0°, 90°, 180°, 270°) and selects the best match.
 * 
 * Common use cases:
 * - Correct images scanned upside-down or sideways
 * - Handle mobile camera captures at any angle
 * - Pre-processing before alignment or detection
 * 
 * Configuration:
 * - referenceImage: Reference marker to match against (Mat)
 * - markerDimensions: Optional resize dimensions for reference [width, height]
 * - threshold: Optional minimum match score requirement
 *   - value: Minimum correlation coefficient (0-1)
 *   - passthrough: If true, continue even if below threshold (with warning)
 */
export class AutoRotate extends Processor {
  private referenceImage: cv.Mat;
  private resizedReference: cv.Mat;
  private threshold?: {
    value: number;
    passthrough: boolean;
  };

  constructor(options: AutoRotateOptions) {
    super();
    this.referenceImage = options.referenceImage;
    this.threshold = options.threshold;

    // Resize reference if marker dimensions specified
    if (options.markerDimensions) {
      const [width, height] = options.markerDimensions;
      this.resizedReference = new cv.Mat();
      cv.resize(
        this.referenceImage,
        this.resizedReference,
        new cv.Size(width, height),
        0,
        0,
        cv.INTER_LINEAR
      );
    } else {
      this.resizedReference = this.referenceImage;
    }
  }

  getName(): string {
    return 'AutoRotate';
  }

  /**
   * Rotate image while maintaining original shape.
   */
  private rotateKeepingShape(image: cv.Mat, rotationCode: number): cv.Mat {
    const originalHeight = image.rows;
    const originalWidth = image.cols;

    // Rotate image
    const rotated = new cv.Mat();
    cv.rotate(image, rotated, rotationCode);

    // If dimensions changed, resize back to original shape
    if (rotated.rows !== originalHeight || rotated.cols !== originalWidth) {
      const resized = new cv.Mat();
      cv.resize(
        rotated,
        resized,
        new cv.Size(originalWidth, originalHeight),
        0,
        0,
        cv.INTER_LINEAR
      );
      rotated.delete();
      return resized;
    }

    return rotated;
  }

  process(context: ProcessingContext): ProcessingContext {
    const { grayImage, coloredImage } = context;

    // Try all 4 rotations
    const rotations = [
      { code: null, label: 'No Rotation' },
      { code: cv.ROTATE_90_CLOCKWISE, label: '90°' },
      { code: cv.ROTATE_180, label: '180°' },
      { code: cv.ROTATE_90_COUNTERCLOCKWISE, label: '270°' },
    ];

    let bestVal = -1;
    let bestRotation: number | null = null;

    // Template matching for each rotation
    for (const { code, label } of rotations) {
      let rotatedImg: cv.Mat;
      let needsCleanup = false;

      if (code === null) {
        rotatedImg = grayImage;
      } else {
        rotatedImg = this.rotateKeepingShape(grayImage, code);
        needsCleanup = true;
      }

      // Perform template matching
      const result = new cv.Mat();
      cv.matchTemplate(
        rotatedImg,
        this.resizedReference,
        result,
        cv.TM_CCOEFF_NORMED
      );

      const minMax = cv.minMaxLoc(result, new cv.Mat());
      const maxVal = minMax.maxVal;

      console.debug(
        `AutoRotate template matching: ${label} = ${maxVal.toFixed(4)}`
      );

      result.delete();
      if (needsCleanup) {
        rotatedImg.delete();
      }

      if (maxVal > bestVal) {
        bestVal = maxVal;
        bestRotation = code;
      }
    }

    // Check threshold if specified
    if (this.threshold && this.threshold.value > bestVal) {
      const message = `AutoRotate score ${bestVal.toFixed(4)} below threshold ${this.threshold.value}`;
      
      if (this.threshold.passthrough) {
        console.warn(`${message}. Continuing due to passthrough flag.`);
      } else {
        throw new Error(
          `${message}. Adjust threshold or check reference marker and input image.`
        );
      }
    }

    console.info(
      `AutoRotate applied: rotation=${bestRotation === null ? 'none' : rotations.find(r => r.code === bestRotation)?.label}, score=${bestVal.toFixed(4)}`
    );

    // Apply best rotation if needed
    if (bestRotation !== null) {
      const rotatedGray = this.rotateKeepingShape(grayImage, bestRotation);
      grayImage.delete();
      context.grayImage = rotatedGray;

      // Rotate colored image if present
      if (coloredImage && !coloredImage.empty()) {
        const rotatedColor = this.rotateKeepingShape(
          coloredImage,
          bestRotation
        );
        coloredImage.delete();
        context.coloredImage = rotatedColor;
      }
    }

    return context;
  }

  /**
   * Clean up allocated resources.
   */
  cleanup(): void {
    if (this.resizedReference !== this.referenceImage) {
      this.resizedReference.delete();
    }
  }
}
